content/blog/waddlesplash/2023-04-24_condition_variables.md

```diff
@@ -0,0 +1,416 @@
+
+Now, the member variables of `ConditionVariableEntry`:
+```cpp
+			ConditionVariable*	fVariable;
+			Thread*				fThread;
+			status_t			fWaitStatus;
+```
+The first is, of course, a pointer to the `Variable` this `Entry` has been added to (or `NULL` if it has not been added to one; or if it was added, but has since been removed.) `fVariable` being non-`NULL` *always* implies that the `Variable` in question also has this `Entry` in its `fEntries` list; the two are never changed except in tandem.
+
+The second is a pointer to the kernel `Thread` structure, which indicates the thread that this `Entry` belongs to, which will wait on it. The `Variable` uses this to wake up the thread in question, but it is also used to signal (by either the `Entry` itself or the `Variable`) that the `Entry` is in the process of being removed from the `Variable`. One principle that is important to note is that whichever party unsets `fThread` is also the same party responsible for decrementing `fEntriesCount`, as we will see later on.
+
+Finally, we have `fWaitStatus`. This value is either positive (if the `Entry` has been `Add`ed somewhere, but not yet notified and woken up), or `<= 0` (once it has been.) When `Wait` is invoked, it will either return this value or a timeout-related error code, depending on what happens. It has some relevance for synchronization, but not very much, as we will (once again) see later on.
+
+#### `Add`ing `Entry` to a `Variable`
+
+This is the most straightforward action of all: the `Variable`’s `fLock` is acquired, and then the following actions are taken inside the `Entry`:
+```
+inline void
+ConditionVariableEntry::_AddToLockedVariable(ConditionVariable* variable)
+{
+	ASSERT(fVariable == NULL);
+
+	fThread = thread_get_current_thread();
+	fVariable = variable;
+	fWaitStatus = STATUS_ADDED;
+	fVariable->fEntries.Add(this);
+	atomic_add(&fVariable->fEntriesCount, 1);
+}
+```
+The `fThread` is set to the one we are currently running on, the `fWaitStatus` is set to `STATUS_ADDED`, we append ourselves to the variable’s `fEntries`, and the `fEntriesCount` is incremented.
+
+#### An `Entry` removing itself from a `Variable`
+
+If an `Entry` is destroyed without having been woken up by the `Variable` it was added to, it must remove itself from that `Variable` before it becomes invalid. Thus, in `~ConditionVariableEntry()`, if `fVariable != NULL`, a method is invoked to perform the removal (which is also used in another condition, as we will see later.)
+
+It is possible that, while the `Entry` is trying to remove itself from the `Variable`, that the `Variable` is also trying to notify and remove this very same `Entry`. This section is thus considered “critical”, and so we begin by entering interrupts-disabled mode, even before we try to acquire the `Variable`’s spinlock.
+
+```cpp
+void
+ConditionVariableEntry::_RemoveFromVariable()
+{
+	InterruptsLocker _;
```

Comment:
this syntax deserves some explanation for people not used to C++ code and our strange naming convention for these variables?